Report special events as check results

therazix opened the pull request:
This PR adds new `internal` checks that run for every test to report special events like timeouts or aborts.

Resolves: #3552 

Pull Request Checklist

* [x] implement the feature
* [ ] write the documentation
* [x] extend the test coverage
* [ ] update the specification
* [ ] adjust plugin docstring
* [ ] modify the json schema
* [ ] mention the version
* [ ] include a release note

happz commented on tmt/checks/internal/timeout.py:
```diff
@@ -0,0 +1,53 @@
+
+
+@provides_check(CHECK_NAME)
+class InternalTimeout(CheckPlugin[InternalTimeoutCheck]):
```
I believe you don't need to include `Internal` in every class name, as they are already "marked" as internal by being in the `tmt.checks.internal` package.

happz commented on tmt/checks/__init__.py:
```diff
@@ -223,6 +223,26 @@ def delegate(
+        for check_plugin in _CHECK_PLUGIN_REGISTRY.iter_plugins():
+            plugin = cast(CheckPlugin[CheckT], check_plugin)
+            if issubclass(plugin._check_class, InternalCheck):
+                checks.append(cast(InternalCheck, plugin._check_class).create_default(logger))
```
Probably rely on the duck typing, with `hasmethod(plugin._check_class, 'create_default')`. Or, add the same method to `Check` and let it return nothing, `InternalCheck` would then provide its own implementation.

therazix commented on tmt/checks/internal/permission.py:
```diff
@@ -0,0 +1,53 @@
+from typing import Optional
```
Renamed in 313ff5e98fb92a586487913171095fe21e949c94.

martinhoyer reviewed (approved):
LGTM.
There seems to be quite a lot of duplication, but I assume trying to improve that would run into the issue of checks having only it's "name" as an identefier? @happz  

Not sure how the test failure is related:
`/tests/execute/framework/shell*`
```
/notexec:
  summary: Testing file permission denied
  test: /dev/null
```
```
    testName="/tests/notexec"
    rlPhaseStartTest "${testName}"
        rlRun -s "${tmt_command} ${testName} >/dev/null" 2 "Testing shell file not executable error"
        rlAssertGrep "cmd: /dev/null" $rlRun_LOG
        rlAssertGrep "/dev/null: Permission denied" $rlRun_LOG
        rlAssertGrep "errr /tests/notexec" $rlRun_LOG
        rlRun -s "${extract_results_command} ${run}/plans/execute/results.yaml"
        rlAssertGrep "/tests/notexec 1 error default-0 null" $rlRun_LOG
    rlPhaseEnd
```

martinhoyer commented:
@coderabbitai full review

happz commented on tmt/checks/internal/abort.py:
```diff
@@ -0,0 +1,55 @@
+        environment: Optional[tmt.utils.Environment] = None,
+        logger: tmt.log.Logger,
+    ) -> list[CheckResult]:
+        return []
```
The default implementation should already return an empty list, therefore it should be safe to drop this custom method.

therazix replied:
Removed in 3a4ba521c2a13eaa7524b06d1e87c24f3c3e9c85.

happz commented on tmt/checks/internal/timeout.py:
```diff
@@ -0,0 +1,61 @@
+        environment: Optional[tmt.utils.Environment] = None,
+        logger: tmt.log.Logger,
+    ) -> list[CheckResult]:
+        return []
```
The default implementation should already return an empty list, therefore it should be safe to drop this custom method.

therazix replied:
Removed in 3a4ba521c2a13eaa7524b06d1e87c24f3c3e9c85.

happz commented on tmt/steps/execute/internal.py:
```diff
@@ -568,6 +569,9 @@ def _save_process(
                 invocation.process = process
 
+        # Insert internal checks
+        invocation.test.check += CheckPlugin.internal_checks(logger)
```
I guess so, yes. Ok, let's leave it here, unless we find a better place.

therazix commented on tmt/checks/internal/permission.py:
```diff
@@ -0,0 +1,61 @@
+        environment: Optional[tmt.utils.Environment] = None,
+        logger: tmt.log.Logger,
+    ) -> list[CheckResult]:
+        return []
```
Removed in 3a4ba521c2a13eaa7524b06d1e87c24f3c3e9c85.

happz commented:
> Thanks for implementing this! Looks nice. Added a few comments. Also, when checking the exceeded timeout scenario I've noticed the `timeout` error is now reported three times:
> 
> ```
> /plan
>     report
>         how: display
>             00:00:05 errr /test/long/beakerlib
>                 Note: timeout
>                 Note: check 'internal/timeout' failed
>                 00:00:00 fail internal/timeout (after-test check)
>                     Note: Test exceeded maximum duration of 5s
>             00:00:05 errr /test/long/shell
>                 Note: timeout
>                 Note: check 'internal/timeout' failed
>                 00:00:00 fail internal/timeout (after-test check)
>                     Note: Test exceeded maximum duration of 5s
>             00:00:02 pass /test/short/beakerlib
>             00:00:01 pass /test/short/shell
>         summary: 2 tests passed and 2 errors
> ```
> 
> Do we really want this? Or shall we drop one of the notes?

Probably, we will end up with 2:

* the `timeout` note is added by the framework - I'd say that should be removed in this PR, because now these special checks take over, and frameworks do not need to be bothered with neither timeouts nor PID file issues.
* the `check 'internal/timeout' failed` note is there because a check has failed, just like `check 'avc' failed` note would appear, I wouldn't remove this one.
* the `Note: Test exceeded maximum duration of 5s` is the more detailed note attached to the check, which should also stay.

So it seems like we will have 2 notes anyway, 1 because of "a check failed" condition, and one actual note.

therazix commented on tests/test/check/test-internal.sh:
```diff
@@ -0,0 +1,57 @@
+    rlPhaseEnd
+
+    rlPhaseStartTest "Passing test with $PROVISION_HOW"
+        rlRun "tmt run --id $run --scratch -a -vv provision -h $PROVISION_HOW test -n /internal/pass"
```
Added in 19d29b0af20e2b301659f080c92e48ee095084d4.

therazix commented:
> > Thanks for implementing this! Looks nice. Added a few comments. Also, when checking the exceeded timeout scenario I've noticed the `timeout` error is now reported three times:
> > ```
> > /plan
> >     report
> >         how: display
> >             00:00:05 errr /test/long/beakerlib
> >                 Note: timeout
> >                 Note: check 'internal/timeout' failed
> >                 00:00:00 fail internal/timeout (after-test check)
> >                     Note: Test exceeded maximum duration of 5s
> >             00:00:05 errr /test/long/shell
> >                 Note: timeout
> >                 Note: check 'internal/timeout' failed
> >                 00:00:00 fail internal/timeout (after-test check)
> >                     Note: Test exceeded maximum duration of 5s
> >             00:00:02 pass /test/short/beakerlib
> >             00:00:01 pass /test/short/shell
> >         summary: 2 tests passed and 2 errors
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Do we really want this? Or shall we drop one of the notes?
> 
> Probably, we will end up with 2:
> 
>     * the `timeout` note is added by the framework - I'd say that should be removed in this PR, because now these special checks take over, and frameworks do not need to be bothered with neither timeouts nor PID file issues.
> 
>     * the `check 'internal/timeout' failed` note is there because a check has failed, just like `check 'avc' failed` note would appear, I wouldn't remove this one.
> 
>     * the `Note: Test exceeded maximum duration of 5s` is the more detailed note attached to the check, which should also stay.
> 
> 
> So it seems like we will have 2 notes anyway, 1 because of "a check failed" condition, and one actual note.

Removed notes added by framework in 19d29b0af20e2b301659f080c92e48ee095084d4.

psss reviewed (approved):
Thanks for the changes, looks good!
